private/bufpkg/bufimage/bufimagemodify/bufimagemodifyv2/field_option_modifier.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufimagemodifyv2
+
+import (
+	"fmt"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufimage"
+	"github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagemodify/internal"
+	"github.com/bufbuild/protocompile/walk"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+type fieldOptionModifier struct {
+	marker                Marker
+	imageFile             bufimage.ImageFile
+	fieldNameToDescriptor map[string]*descriptorpb.FieldDescriptorProto
+	fieldNameToSourcepath map[string][]int32
+}
+
+func newFieldOptionModifier(
+	imageFile bufimage.ImageFile,
+	marker Marker,
+) (*fieldOptionModifier, error) {
+	fieldNameToDescriptor := make(map[string]*descriptorpb.FieldDescriptorProto)
+	fieldNameToSourcePath := make(map[string][]int32)
+	err := walk.DescriptorProtosWithPath(
+		imageFile.Proto(),
+		func(
+			fullName protoreflect.FullName,
+			path protoreflect.SourcePath,
+			message proto.Message,
+		) error {
+			fieldDescriptor, ok := message.(*descriptorpb.FieldDescriptorProto)
+			if !ok {
+				return nil
+			}
+			fieldNameToDescriptor[string(fullName)] = fieldDescriptor
+			fieldNameToSourcePath[string(fullName)] = path
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &fieldOptionModifier{
+		marker:                marker,
+		imageFile:             imageFile,
+		fieldNameToDescriptor: fieldNameToDescriptor,
+		fieldNameToSourcepath: fieldNameToSourcePath,
+	}, nil
+}
+
+func (m *fieldOptionModifier) FieldNames() []string {
+	fieldNames := make([]string, 0, len(m.fieldNameToDescriptor))
+	for fieldName := range m.fieldNameToDescriptor {
+		fieldNames = append(fieldNames, fieldName)
+	}
+	return fieldNames
+}
+
+func (m *fieldOptionModifier) ModifyJSType(
+	fieldName string,
+	override Override,
+) error {
+	if internal.IsWellKnownType(m.imageFile) {
+		return nil
+	}
```

Comment:
> Do we skip WKT for other managed options as well?

So far, all the other options are file options and we skip them for WKT files.

> Timestamp and Duration have int64 fields.

Great point! I did not consider that and took it for granted that WKT wouldn't be modified.

> Implementations that ship pre-generated WKT and need a different jstype will have to modify themselves.

I'm not sure if I get what you meant by '[having] to modify themselves', could you give me some more details on this? (also I assumed an implementation means a plugin)

Reply:
> Now I do think that it should be modified for WKT as well, unless disabled.

That would be helpful for JS implementations. I can't think of a negative impact if its limited to field options.

Reply:
Talked offline. Specifically for `jstype`, including WKT seems harmless. But in case JS will ever get a file option similar to `go_package`, the behavior will be surprising. Consistency seems more important to me - let's not modify WKT.